assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -1,20 +1,54 @@
+# On platforms like AWS, it's enough for kube-apiserver to listen on Node InternalIP. On other
+# like Packet, we need to listen on all interfaces to be able to expose kube-apiserver to the
+# internet. As we cannot listen on specific IP address and on 0.0.0.0 on the same port when using
+# SO_REUSEPORT, kube-apiserver process must listen on different port than HAProxy does. So if we
+# need to expose kube-apiserver on all interfaces, kube-apiserver will be listening on random
+# 127.0.0.0/8 IP address on port 7443, but will advertise Node InternalIP with port 7443 and HAProxy
+# will be listening on both Node InternalIP on port 7443 and on 0.0.0.0:6443 for both internal and
+# external traffic.
+# See commit messages and comments :https://github.com/kinvolk/lokomotive/pull/32 for more details.
+{{ define "port" }}
+{{- if and .Values.apiserver.exposeOnAllInterfaces (eq (int .Values.apiserver.replicas) 1) -}}
+7443
+{{- else -}}
+6443
+{{- end -}}
+{{ end }}
 apiVersion: apps/v1
+# If there is jsut one controller node, we want to use Deployment to be able to run 2 kube-apiserver
+# pods on a single node at a time, to provide graceful upgrades.
+# See commit messages and comments in https://github.com/kinvolk/lokomotive/pull/32 for more details.
```

Comment:
Again, I'm not a fan of the "go and spend 10 hours reading" instead of providing a link to the specific relevant comment

Reply:
I don't know where to point there, as there is to many aspects to take into account. Maybe I should have a design doc describing that? :joy: CC @iaguis

Reply:
What about pointing to this commit: https://github.com/kinvolk/lokomotive/pull/32/commits/04f9737a4947d89e51f9ae066a319c92aaa5c27a (of course, hash might change on ammends, just update it as a very last thing?)

I think there explains the problem in detail quite well

Reply:
Ok, will do.

Reply:
Done. I hope I don't forget to update it before merging, if there are further changes :see_no_evil: